gtsfm/densify/mvs_patchmatchnet.py

```diff
@@ -0,0 +1,251 @@
+
+Authors: Ren Liu
+"""
+import copy
+import time
+from typing import Dict, List
+
+import numpy as np
+import torch
+import torch.nn as nn
+from torch.utils.data import DataLoader
+
+from gtsfm.common.image import Image
+from gtsfm.common.gtsfm_data import GtsfmData
+from gtsfm.densify.mvs_base import MVSBase
+from gtsfm.densify.patchmatchnet_data import PatchmatchNetData
+from gtsfm.utils import logger as logger_utils
+from thirdparty.patchmatchnet.models.net import PatchmatchNet
+from thirdparty.patchmatchnet.utils import tensor2numpy, tocuda
+from thirdparty.patchmatchnet.eval import check_geometric_consistency
+
+logger = logger_utils.get_logger()
+
+torch.backends.cudnn.benchmark = True
+PATCHMATCHNET_WEIGHTS_PATH = "thirdparty/patchmatchnet/checkpoints/model_000007.ckpt"
+DEFAULT_VIEW_NUMBER = 5
+DEFAULT_GEOMETRIC_PIXEL_THRESH = 1.0
+DEFAULT_GEOMETRIC_DEPTH_THRESH = 0.01
+DEFAULT_PHOTOMETRIC_THRESH = 0.8
+DEFAULT_BATCH_SIZE = 1
+DEFAULT_WORKERS_NUMBER = 4
+
+
+class MVSPatchmatchNet(MVSBase):
+    """Inference methods for GtsfmData in densification using PatchmatchNet."""
+
+    def densify(
+        self,
+        images: Dict[int, Image],
+        sfm_result: GtsfmData,
+        num_views: int = DEFAULT_VIEW_NUMBER,
+        thresholds: List[float] = [
+            DEFAULT_GEOMETRIC_PIXEL_THRESH,
+            DEFAULT_GEOMETRIC_DEPTH_THRESH,
+            DEFAULT_PHOTOMETRIC_THRESH,
+        ],
+    ) -> np.ndarray:
+        """Get dense point cloud using PatchmatchNet from GtsfmData. The method implements the densify method in MVSBase
+        Ref: Wang et al. https://github.com/FangjinhuaWang/PatchmatchNet/blob/main/eval.py
+
+        Args:
+            images: image dictionary obtained from loaders
+            sfm_result: result of GTSFM after bundle adjustment
+            num_views: number of views, containing 1 reference view and (num_views-1) source views
+                Defaults to DEFAULT_VIEW_NUMBER
+            thresholds: geometric pixel threshold, geometric depth threshold, and photometric
+                threshold for filtering inference results.
+                Defaults to [DEFAULT_GEOMETRIC_PIXEL_THRESH, DEFAULT_GEOMETRIC_DEPTH_THRESH, DEFAULT_PHOTOMETRIC_THRESH]
+                1. for geometric thresholds, small threshold means high accuracy and low completeness
+                2. for photometric thresholds, large threshold means high accuracy and low completeness
+
+        Returns:
+            3D coordinates (in the world frame) of the dense point cloud, (point number, 3)
+        """
+        dataset = PatchmatchNetData(images=images, sfm_result=sfm_result, num_views=num_views)
+        loader = DataLoader(
+            dataset=dataset,
+            batch_size=DEFAULT_BATCH_SIZE,
+            shuffle=False,
+            num_workers=DEFAULT_WORKERS_NUMBER,
+            drop_last=False,
+        )
+
+        model = PatchmatchNet(
+            patchmatch_interval_scale=[0.005, 0.0125, 0.025],
```

Comment:
nit: these seem like magic numbers / defaults that should go at the top of the file

Reply:
resolved, thanks